### ISSUE TITLE
[feature] Enable/disable event listening on @HostListeners

### ISSUE THREAD

**Issue by adamdbradley** [angular/angular#7626]
There are cases where `@HostListener` is the ideal API to use, however, for events like `touchmove`, `mousemove` or `scroll` we cannot use them for performance reasons. Far too many and often unnecessary change detections are happening. Additionally, there are many times we don't even need to listen in on `touchmove` events at all, unless a certain criteria happens. 

One example is when an input has focus, we want to know when the user begins to scroll it's parent, however, we don't need every input to listen to every scroll event. The other is for our pull-to-refresh and infinite scroll features, where users may or may not have them enabled. When they're not enabled at all, there's no need to still be listening in on every touchmove/scroll events.

The current work around is for us to manually add and remove touchmove or scroll listeners depending on our criteria. While this works, we need to be very careful about not adding memory leaks, but more importantly we're not using `@HostListener` at all which keeps code nice and clean.

Could a feature be added to enable/disable host listeners from listening to the actual native events? This isn't about not firing off events under a certain criteria, but completely removing native event listeners when they're not needed.

Discussed this with both @IgorMinar and @mhevery, and said this would also be related to work needed for https://github.com/angular/angular/issues/1773

**Comment by robwormald** [comment]
another way to handle this (would require some work) would be Observables - given they are cold and do not do register the event handlers unless they are actually subscribed to, there would be no overhead to simply creating them..... related to #4062

**Comment by esugoi** [comment]
+1
This would be really great. :)
Just enabling\disabling host listeners would already be a great help. There are many cases where I am using the Renderer interface to handle\unhandle DOM events as a work-around, since I can't do this using only the host listeners.

Obviously if we could use expressions\criterias too would be even greater. ;)

**Comment by pkozlowski-opensource** [comment]
I could use some solution for bootstrap components as well....

**Comment by zoechi** [comment]
related to or dup of #1773 ?

**Comment by pkozlowski-opensource** [comment]
Yes, this is essentially the same as #1773, going to close as a duplicate. @adamdbradley hope you don't mind, let's continue the discussion in the #1773

**Comment by adamdbradley** [comment]
@pkozlowski-opensource This isn't about debouncing, but rather not having listeners added to the element at all, and ability to add/remove those listeners.

**Comment by zoechi** [comment]
If you want to control when these events are subscribed/unsubscribed to, then why not just subscribe imperatively?

**Comment by naveedahmed1** [comment]
Any update on this?

**Comment by GiuseppePiscopo** [comment]
@zoechi I take OP wish to keep using `@HostListener` for its (declarative) terseness, as opposed to listening and unlistening to/from event imperatively.

@pkozlowski-opensource I too think that this issue is not related to debouncing o rate-reduction. This is about turning listener on and off (and on again?) when something happens. Something like (just some unexistent syntax):

```ts
@HostListener('document:mouseup', ['$event'], 'myPredicate')
onDocumentMouseUp(evt: MouseEvent) {
  // ...
}

myPredicate(): boolean {
  // ...
  return shouldListenOrNot;
}
```

@robwormald I'd be interested to know how can you plug an Observable to this HostListener, so that it gets listened/unlistened appropriately. Thanks

I'd like to know as well if there's any progress on this. Thanks everyone.

**Comment by Romchik22** [comment]
The Angular has interesting ability for partly solving this problem. https://angular.io/api/core/Renderer2#listen
` constructor(private renderer: Renderer2)
 let listener = this.renderer.listen('window', 'keydown.esc', (event) => // do some logic);
this.listener() // remove EventListener for example in ngOnDestroy() or in another time when do you need. `

**Comment by waterplea** [comment]
I have created an EventListenerPlugin exactly for that purpose. It works fine coupled with a directive when subscribing in template using `(eventName)="handler($event)"` and it also works for @HostListeners but I would really love to either create my own decorator to define predicate or be able to pass it on as an argument to @HostListener as it is not that intuitive to declare them using service:
https://stackblitz.com/edit/angular-event-filter

Any ideas to help me improve it in that way would be greatly appreciated :)

**Comment by jpduckwo** [comment]
Just leaving this here for interest. In my case I wanted to only listen for the first occurrence of an event then stop listening. This was for detecting if the device is a touch enabled device. The Renderer2 worked well thanks @SobolRR 
```
    ...
    // Listen for the first isTouch event then stop listening
    this.isTouchUnlisten = this.renderer2.listen('window', 'touchstart', () => {
      // Do something to set as a touch device
      this.isTouchUnlisten();
    });
```

**Comment by rdtbert** [comment]
We had an custom input component which would cause a performance issue the more times we added this field type to a page because of a @HostListener. 
 
 @HostListener('window:keydown', ['$event'])
    keyEvent(event: KeyboardEvent) { do special stuff   }
    }

Don't know if this is the best way  but we solved it by dynamically adding and removing the handler during a focus and blur.
   <mat-select formControlName="stop"
        (focus)="_handleFocus()"
        (blur)="_handleBlur()" >

_handleFocus(): void {
   // dynamically inserts a HostListener for keydown
   if (! this.keyDownlistener) 
      this.keyDownlistener =  this.renderer.listen(window, "keydown", event =>{ 
         --do special stuff--
        }  });   
  }

_handleInput(): void {    
    //remove hostListener for keydown
    if (this.keyDownlistener) {this.keyDownlistener(); this.keyDownlistener = null;};
}

**Comment by mhombach** [comment]
> We had an custom input component which would cause a performance issue the more times we added this field type to a page because of a @HostListener.
> 
> @HostListener('window:keydown', ['$event'])
> keyEvent(event: KeyboardEvent) { do special stuff }
> }
> 
> Don't know if this is the best way but we solved it by dynamically adding and removing the handler during a focus and blur.
> <mat-select formControlName="stop"
> (focus)="_handleFocus()"
> (blur)="_handleBlur()" >
> 
> **//we used Render2**
> _handleFocus(): void {
> // dynamically inserts a HostListener for keydown
> if (! this.keyDownlistener)
> this.keyDownlistener = this.renderer.listen(window, "keydown", event =>{
> --do special stuff--
> } });
> }
> 
> _handleInput(): void {
> //remove hostListener for keydown
> if (this.keyDownlistener) {this.keyDownlistener(); this.keyDownlistener = null;};
> }

Yes. It is always a good practice to use Renderer2 and listen "on demand" for events that should only be listened to by some condition or only in certain scenarios.
The bigger your app get's, the more you should think about performance and you can easily replace ordinary HostListeners by onDemand-listeners (with Renderer2), once you got used to them :)

**Comment by waterplea** [comment]
Hey I gave a link to stackblitz for an EventManagerPlugin I've made a few message above. Since then I reworked this approach to a production-ready level and released it as an npm package. So far it worked great for me, kinda like the onDemand listeners approach mentioned above. Here it is on Github: 
https://github.com/TinkoffCreditSystems/ng-event-plugins
In Demo section of the readme there's a link to detailed article about how it works and an interactive demo. Hope it might be helpful to you. Can't wait for Angular 10 so I can build it with Ivy and add other features like capture phase listeners or passive listeners :)

**Comment by aniket-kale** [comment]
> Just leaving this here for interest. In my case I wanted to only listen for the first occurrence of an event then stop listening. This was for detecting if the device is a touch enabled device. The Renderer2 worked well thanks @SobolRR
> 
> ```
>     ...
>     // Listen for the first isTouch event then stop listening
>     this.isTouchUnlisten = this.renderer2.listen('window', 'touchstart', () => {
>       // Do something to set as a touch device
>       this.isTouchUnlisten();
>     });
> ```

Thanks, works perfectly.

**Comment by Splinteer** [comment]
+1 it would be very nice if we can listen/unlisten from @HostListener.
This would avoid listening for the entire time for no reason and improve performances.
This would also be way more user friendly/cleaner to do it using @HostListener instead of using Renderer2
